packages/grid/_modules_/grid/hooks/features/columnReorder/useGridColumnReorder.tsx

```diff
@@ -103,27 +103,33 @@ export const useGridColumnReorder = (apiRef: GridApiRef): void => {
       const coordinates = { x: event.clientX, y: event.clientY };
 
       if (
-        params.field !== dragCol &&
+        params.field !== dragColField &&
         hasCursorPositionChanged(cursorPosition.current, coordinates)
       ) {
         const targetColIndex = apiRef.current.getColumnIndex(params.field, false);
-        const dragColIndex = apiRef.current.getColumnIndex(dragCol, false);
-
-        if (
-          (getCursorMoveDirectionX(cursorPosition.current, coordinates) ===
-            CURSOR_MOVE_DIRECTION_RIGHT &&
-            dragColIndex < targetColIndex) ||
-          (getCursorMoveDirectionX(cursorPosition.current, coordinates) ===
-            CURSOR_MOVE_DIRECTION_LEFT &&
-            targetColIndex < dragColIndex)
-        ) {
-          apiRef.current.setColumnIndex(dragCol, targetColIndex);
+        const targetColVisibleIndex = apiRef.current.getColumnIndex(params.field, true);
+        const targetCol = apiRef.current.getColumn(params.field);
+        const dragColIndex = apiRef.current.getColumnIndex(dragColField, false);
+        const visibleColumnAmount = apiRef.current.getVisibleColumns().length;
+
+        const canBeReordered =
+          !targetCol.disableReorder ||
+          (targetColVisibleIndex > 0 && targetColVisibleIndex < visibleColumnAmount - 1);
+
+        const cursorMoveDirectionX = getCursorMoveDirectionX(cursorPosition.current, coordinates);
```

Comment:
For another PR:

This `cursorMoveDirectionX` is wrong. It's always "right" the first time. We need to save the cursor position when the user starts dragging